src/schedule/mod.rs

```diff
@@ -92,95 +162,126 @@ impl JoinerSecret {
-                key_package_hash,
-            });
+pub(crate) struct KeySchedule {
+    ciphersuite: &'static Ciphersuite,
+    intermediate_secret: Option<IntermediateSecret>,
+    epoch_secret: Option<EpochSecret>,
+    state: State,
+}
+
+impl KeySchedule {
+    /// Initialize the key schedule and return it.
+    pub(crate) fn init(
+        ciphersuite: &'static Ciphersuite,
+        joiner_secret: JoinerSecret,
+        psk: impl Into<Option<Secret>>,
+    ) -> Self {
+        let intermediate_secret = IntermediateSecret::new(ciphersuite, &joiner_secret, psk.into());
+        Self {
+            ciphersuite,
+            intermediate_secret: Some(intermediate_secret),
+            epoch_secret: None,
+            state: State::Initial,
         }
-        Ok(plaintext_secrets)
     }
-}
 
-pub(crate) struct PlaintextSecret {
-    pub(crate) public_key: HPKEPublicKey,
-    pub(crate) group_secrets_bytes: Vec<u8>,
-    pub(crate) key_package_hash: Vec<u8>,
+    /// Derive the welcome secret.
+    /// Note that this has to be called before the context is added.
+    pub(crate) fn welcome(&self) -> Result<WelcomeSecret, KeyScheduleError> {
+        if self.state != State::Initial || self.intermediate_secret.is_none() {
+            log::error!("Trying to derive a welcome secret while not in the initial state.");
+            return Err(KeyScheduleError::InvalidState(ErrorState::NotInit));
+        }
+
+        Ok(WelcomeSecret::new(
+            self.ciphersuite,
+            self.intermediate_secret.as_ref().unwrap(),
+        ))
+    }
+
+    /// Add the group context to the key schedule.
+    pub(crate) fn add_context(
+        &mut self,
+        group_context: &GroupContext,
+    ) -> Result<(), KeyScheduleError> {
+        if self.state != State::Initial || self.intermediate_secret.is_none() {
+            log::error!(
+                "Trying to add context to the key schedule while not in the initial state."
+            );
+            return Err(KeyScheduleError::InvalidState(ErrorState::NotInit));
+        }
+        self.state = State::Context;
+
+        self.epoch_secret = Some(EpochSecret::new(
+            self.ciphersuite,
+            self.intermediate_secret.as_ref().unwrap(),
+            group_context,
+        ));
+        self.intermediate_secret = None;
+        Ok(())
+    }
+
+    /// Derive the init secret.
+    pub(crate) fn init_secret(&mut self) -> Result<InitSecret, KeyScheduleError> {
+        if self.state != State::Context || self.epoch_secret.is_none() {
+            log::error!("Trying to derive the init secret while not in the initial state.");
+            return Err(KeyScheduleError::InvalidState(ErrorState::NotContext));
+        }
+
+        Ok(InitSecret::new(
+            self.ciphersuite,
+            self.epoch_secret.as_ref().unwrap(),
+        ))
+    }
+
+    /// Derive the epoch secrets.
+    pub(crate) fn epoch_secrets(&mut self) -> Result<EpochSecrets, KeyScheduleError> {
```

Comment:
There is no order to these secrets. There are only two different functions because we don't always need the `init_secret`. If we want to get the additional guarantee that we can't derive secrets multiple times, we can do so with adding more complexity here. I added the additional state.
But note that first deriving the `EpochSecrets` and then the `InitSecrets` is the right order if the state machine should terminate. We don't always derive the `init_secret`, which would leave the state machine in an open state.
The other option to solve this is integrating the `init_secret` into the `EpochSecrets` struct and only deriving it when requested.

Reply:
An alternative could be modelling the key schedule as a state machine with a loop, representing the chain that it is in the spec. That would mean it permanently lives in the group state, but that could be a viable option.

Reply:
From all secrets, the init secret is the most sensitive one, because if it leaks it allows an attacker to break into a group permanently. That's why I am concerned about it being derived more than once. 
You have a point about the fact, that we don't always derive the init secret though. I personally don't care to much about the state machine terminating, being able to derive the init secret only once is more important to me. If we want the state machine to always terminate I guess we would need two different functions:
1. Derive the epoch secrets and terminate
2. Derive the epoch secrets, then the init secret and then terminate.

Reply:
Ok, I don't actually think this is super important. Deleting the key material is more important. But the cleanest option then is to do it all at the same time (the `init_secret` optionally). I'll do that.